bin/citgm-all

```diff
@@ -102,93 +101,39 @@ function runCitgm (mod, name, next) {
-      }
-      else {
-        exitCode = 1;
-        failed.push(module);
-      }
+  }).on('end', function(result) {
+    if (result.error) {
+      log.error('done', 'The test suite for ' + result.name + ' version ' + result.version + ' failed');
     }
     else {
-      passed.push(module);
-      log.info('done', 'The test suite for ' + module.name + ' version ' + module.version + ' passed.');
+      log.info('done', 'The test suite for ' + result.name + ' version ' + result.version + ' passed.');
     }
-    next();
+    result.flaky = mod.flaky;
+    modules.push(result);
+    next(null);
   }).run();
 }
 
-function printModule(logType, module) {
-  log[logType](chalk.yellow('module name:'), module.name);
-  log[logType](chalk.yellow('version:'), module.version);
-  if (module.error) {
-    log[logType]('error', module.error.message);
-  }
-}
-
-function printModules(logType, modules) {
-  _.each(modules, function (module) {
-    printModule(logType, module);
-  });
-}
-
-function printModulesMarkdown(title, modules) {
-  /*eslint-disable no-console*/
-  if (modules.length > 0) {
-    console.log('### ' + title + ' Modules');
-    _.each(modules, function (module) {
-      console.log('* ' + module.name);
-      console.log('  - version ' + module.version);
-      if (module.error) {
-        console.log('  * ' + module.error.message);
-      }
-    });
-    console.log('');
-  }
-  /*eslint-enable no-console*/
-}
-
-function printMarkdown() {
-  /*eslint-disable no-console*/
-  if (failed.length === 0) {
-    console.log('## 🎉🎉 CITGM Passed 🎉🎉');
-  }
-  else {
-    console.log('## 🔥⚠️🔥 CITGM FAILED 🔥⚠️🔥');
-  }
-  /*eslint-enable no-console*/
-  printModulesMarkdown('Passing', passed);
-  printModulesMarkdown('Flaky', flaky);
-  printModulesMarkdown('Failing', failed);
-}
-
 function launch() {
-  async.forEachOfSeries(lookup, runCitgm, function done () {
-    if (passed.length > 0) {
-      log.info('passing modules');
-      printModules('info', passed);
-    }
-    if (flaky.length > 0) {
-      log.warn('flaky modules');
-      printModules('warn', flaky);
-    }
-    if (failed.length > 0) {
-      process.exitCode = exitCode;
-      log.error('failing modules');
-      printModules('error', failed);
-      log.error('done', 'citgm-all failed');
-    }
-    else {
-      log.info('done', 'citgm-all passed.');
+  async.forEachOfSeries(lookup, runCitgm, function done (err) {
+    if (err) {
+      log.error('citgm-all failure:', err.message);
     }
+
+    reporter.logger(log, modules);
+
     if (app.markdown) {
-      printMarkdown();
+      reporter.markdown(log.bypass, modules);
+    }
+    
+    if (app.tap) {
+      // if tap is a string it should be a path to write output to
+      // if not use `log.bypass` which is currently process.stdout.write
+      // TODO check that we can write to that path, perhaps require a flag to overwrite
+      let tap = (typeof app.tap === 'string') ? app.tap : log.bypass;
```

Comment:
The idea here was that the reporter simply gets a function that has the signature `fn(msg)` which makes it simple to stub in different reporter options. I'm also not 100% that "bypass" is going to survive so it was a quick way to isolate